src/clr-angular/forms/datalist/datalist-container.ts

```diff
@@ -0,0 +1,112 @@
+import { DynamicWrapper } from '../../../clr-angular/utils/host-wrapping/dynamic-wrapper';
+import { DatalistControllerService } from './providers/datalist-controller.service';
+import { ControlIdService } from '../common/providers/control-id.service';
+import { FocusService } from '../common/providers/focus.service';
+import { IfErrorService } from '../common/if-error/if-error.service';
+import { NgControlService } from '../common/providers/ng-control.service';
+import { NgControl } from '@angular/forms';
+import { Subscription } from 'rxjs';
+import { ClrLabel } from '../common/label';
+import { UNIQUE_ID, UNIQUE_ID_PROVIDER } from '../../utils/id-generator/id-generator.service';
+
+@Component({
+  selector: 'clr-datalist-container',
+  template: `
+    <ng-content select="label"></ng-content>
+    <label *ngIf="!label && addGrid()"></label>
+    <div class="clr-control-container" [ngClass]="controlClass()">
+      <div class="clr-input-wrapper">
+        <div class="clr-input-group" [class.clr-focus]="focus">
+          <ng-content select="[clrDatalist]"></ng-content>
+          <datalist [id]="listId">
+            <option *ngFor="let item of items" [value]="item"></option>
+          </datalist>
+        </div>
+        <clr-icon *ngIf="invalid" class="clr-validate-icon" shape="exclamation-circle" aria-hidden="true"></clr-icon>
+      </div>
+      <ng-content select="clr-control-helper" *ngIf="!invalid"></ng-content>
+      <ng-content select="clr-control-error" *ngIf="invalid"></ng-content>
+    </div>
+    `,
+  host: {
+    '[class.clr-form-control]': 'true',
+    '[class.clr-form-control-disabled]': 'control?.disabled',
+    '[class.clr-row]': 'addGrid()',
+  },
+  providers: [
+    ControlClassService,
+    LayoutService,
+    ControlIdService,
+    DatalistControllerService,
+    FocusService,
+    IfErrorService,
+    NgControlService,
+    UNIQUE_ID_PROVIDER,
+  ],
+})
+export class ClrDatalistContainer implements DynamicWrapper {
+  private subscriptions: Subscription[] = [];
+  _dynamic: boolean = false;
+  invalid: boolean = false;
+  focus: boolean = false;
+  control: NgControl;
+  items: string[];
+  listId: string;
+
+  constructor(
+    private datalistControllerService: DatalistControllerService,
+    private controlClassService: ControlClassService,
+    private layoutService: LayoutService,
+    private ifErrorService: IfErrorService,
+    private focusService: FocusService,
+    private ngControlService: NgControlService,
+    @Inject(UNIQUE_ID) public id: string
+  ) {
+    this.subscriptions.push(
+      this.ifErrorService.statusChanges.subscribe(invalid => {
+        this.invalid = invalid;
+      })
+    );
+    this.subscriptions.push(
+      this.focusService.focusChange.subscribe(state => {
+        this.focus = state;
+      })
+    );
+    this.subscriptions.push(
+      this.ngControlService.controlChanges.subscribe(control => {
+        this.control = control;
+      })
+    );
```

Comment:
I believe you can clean this up a bit by combining the subscriptions into the subscription push call.

```typescript
  this.subscriptions.push(
    this.ifErrorService.statusChanges.subscribe(invalid => this.invalid = invalid),
    this.focusService.focusChange.subscribe(state => this.focus = state),
    this.ngControlService.controlChanges.subscribe(control => this.control = control)
  );
```